crates/services/p2p/src/heartbeat/handler.rs

```diff
@@ -0,0 +1,330 @@
+            Some(OutboundState::RequestingBlockHeight {
+                requested: true,
+                stream,
+            }) => {
+                // start new send timeout
+                self.timer.reset(self.config.send_timeout);
+                // send latest `BlockHeight`
+                self.outbound = Some(OutboundState::SendingBlockHeight(
+                    send_block_height(stream, block_height).boxed(),
+                ))
+            }
+            other_state => self.outbound = other_state,
+        }
+    }
+
+    fn poll(
+        &mut self,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<
+        ConnectionHandlerEvent<
+            Self::OutboundProtocol,
+            Self::OutboundOpenInfo,
+            Self::OutEvent,
+            Self::Error,
+        >,
+    > {
+        if let Some(inbound_block_height) = self.inbound.as_mut() {
+            match inbound_block_height.poll_unpin(cx) {
+                Poll::Ready(Err(_)) => {
+                    debug!(target: "fuel-libp2p", "Incoming heartbeat errored");
+                    self.inbound = None;
+                }
+                Poll::Ready(Ok((stream, block_height))) => {
+                    // start waiting for the next `BlockHeight`
+                    self.inbound = Some(receive_block_height(stream).boxed());
+
+                    // report newly received `BlockHeight` to the Behaviour
+                    return Poll::Ready(ConnectionHandlerEvent::Custom(
+                        HeartbeatOutEvent::BlockHeight(block_height),
+                    ))
+                }
+                _ => {}
+            }
+        }
+
+        loop {
+            if self.failure_count >= self.config.max_failures.into() {
+                // Request from `Swarm` to close the faulty connection
+                return Poll::Ready(ConnectionHandlerEvent::Close(
+                    HeartbeatFailure::Timeout,
+                ))
+            }
+
+            match self.outbound.take() {
+                Some(OutboundState::RequestingBlockHeight { requested, stream }) => {
+                    self.outbound = Some(OutboundState::RequestingBlockHeight {
+                        stream,
+                        requested: true,
+                    });
+
+                    if !requested {
+                        return Poll::Ready(ConnectionHandlerEvent::Custom(
+                            HeartbeatOutEvent::RequestBlockHeight,
+                        ))
+                    }
+
+                    break
+                }
+                Some(OutboundState::SendingBlockHeight(mut outbound_block_height)) => {
+                    match outbound_block_height.poll_unpin(cx) {
+                        Poll::Pending => {
+                            if self.timer.poll_unpin(cx).is_ready() {
+                                // Time for successfull send expired!
+                                self.failure_count += 1;
```

Comment:
Do we need to reset the timer here? Otherwise it seems like this could get polled 5x in a row immediately once the timer expires, which is effectively the same as having a failure count of 1.

Reply:
We don't have to, since we have _taken_ the `Option<OutboundState>`, let's say it was
`Some(OutboundState::SendingBlockHeight)`
we check for the `timer` it expired -> we increment the failure count,
next `poll` or `loop` iteration if previous was expired, now the `Option<Outbound>` will be `None`
so it won't hit the timer check again.

The timer is restarted once we move to `OutboundState::SendingBlockHeight` again, [here](https://github.com/FuelLabs/fuel-core/pull/911/files#diff-81c845fee50fec4b1c2cab23105e502b561fec4db6dab63ea158f4d89679c2edR136).